File: src/java/magmac/app/compile/rule/NodeRule.java
package magmac.app.compile.rule;

import magmac.app.compile.error.CompileResult;
import magmac.app.compile.node.MapNode;
import magmac.app.compile.node.Node;

public record NodeRule(String key, Rule childRule) implements Rule {
    @Override
    public CompileResult<Node> lex(String input) {
        return this.childRule.lex(input)
                .mapValue((Node lexed) -> new MapNode().withNode(this.key, lexed));
    }

    @Override
    public CompileResult<String> generate(Node node) {
        return node.findNodeOrError(this.key).flatMapValue((Node child) -> this.childRule.generate(child));
    }
}

File: src/web/magmac/app/compile/rule/NodeRule.ts
import { CompileResult } from "../../../../magmac/app/compile/error/CompileResult";
import { MapNode } from "../../../../magmac/app/compile/node/MapNode";
import { Node } from "../../../../magmac/app/compile/node/Node";
export class NodeRule {
	lex(input : String) : CompileResult<Node> {return 0.childRule.lex( 0).mapValue( 0);;}
	generate(node : Node) : CompileResult<String> {return 0.findNodeOrError( 0.key).flatMapValue( 0.childRule.generate);;}
}
